sprokit/processes/core/throttle_image_stream_process.cxx

```diff
@@ -0,0 +1,174 @@
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "throttle_image_stream_process.h"
+#include <vital/types/image_container.h>
+#include <vital/types/timestamp_config.h>
+#include <vital/vital_types.h>
+#include <vital/types/timestamp.h>
+
+#include <sprokit/processes/kwiver_type_traits.h>
+#include <sprokit/pipeline/process_exception.h>
+
+namespace kwiver {
+
+create_config_trait( rate, double, "1", "Desired rate (Hz) of the output image "
+                     "stream." );
+
+//----------------------------------------------------------------
+// Private implementation class
+class throttle_image_stream_process::priv
+{
+public:
+  double m_tstep {1};
+  double m_next_time;
+  bool m_initialized {false};
+  priv();
+  ~priv();
+
+}; // end priv class
+
+
+// ==================================================================
+throttle_image_stream_process::
+throttle_image_stream_process( kwiver::vital::config_block_sptr const& config )
+  : process( config ),
+    d( new throttle_image_stream_process::priv )
+{
+
+  make_ports();
+  make_config();
+}
+
+
+throttle_image_stream_process::
+~throttle_image_stream_process()
+{
+}
+
+
+// ------------------------------------------------------------------
+void
+throttle_image_stream_process::
+_configure()
+{
+  double rate = config_value_using_trait( rate );
+
+  if( rate <= 0)
+  {
+    throw sprokit::invalid_configuration_exception( name(), "Rate must be greater than zero." );
+  }
+
+  d->m_tstep = 1/rate;
+}
+
+
+// ------------------------------------------------------------------
+void
+throttle_image_stream_process::
+_step()
+{
+  kwiver::vital::timestamp frame_time = grab_input_using_trait( timestamp );
+
+  if( !d->m_initialized )
+  {
+    vital::image_container_sptr input = grab_from_port_using_trait( image );
```

Comment:
This line could be pulled up to go with the grab ror the timestamp, since it is done in all three cases below.